### PR TITLE
fix(deps): request consumers to install react-native polyfills

### PR DESCRIPTION
**Pull request by trivikr**
Co-authored-by: Nicolas Perraut <n.perraut@gmail.com>

### Issue
Related to #1536
Replaces https://github.com/aws/aws-sdk-js-v3/pull/2108

Fixes: https://github.com/aws/aws-sdk-js-v3/issues/1797
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2051
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2192
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2199
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2219

### Description
This PR marks react-native-get-random-values as optional peer dependency to prevent install on non React Native environments.

### Testing
Verified that ReactNative E2E tests are successful.

<details>
<summary>Command line output</summary>

```console
$ pwd
/Users/trivikr/workspace/aws-sdk-js-v3/tests/react-native/End2End

$ node launch-app.js --local-publish
...
 PASS  e2e/End2EndTest.spec.js (14.5s)
  S3
    ✓ should upload an object after tapping putObject button (2276ms)
    ✓ after tapping getObject, App should get the same object that uploaded (2224ms)
    ✓ should list Objects after tapping listObjects button (1496ms)
    ✓ should conduct multipart upload after tapping createMultipartUpload, uploadPart, completeMultipartUpload buttons (3096ms)
```

</details>

<details>
<summary>Details</summary>

https://user-images.githubusercontent.com/16024985/112871068-bb1fe300-9073-11eb-9d25-13d0a2f67ab0.mov

</details>

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by G-Rath** [comment]
@trivikr have you tested this to confirm it doesn't throw an error on import if the package isn't installed?

I would expect it to, which is why we used a try/catch over in #2108

**Comment by trivikr** [comment]
> @trivikr have you tested this to confirm it doesn't throw an error on import if the package isn't installed?

The AWS SDK for JavaScript no longer imports react-native polyfills, and expect them to be imported by the app using the SDK in react-native environment.

I'll evaluate the behavior if the react-native app doesn't contain react-native polyfills, and update this PR.

**Comment by G-Rath** [comment]
Ah right I see - sorry it didn't click that [this change](https://github.com/aws/aws-sdk-js-v3/pull/2108/files#diff-97aa9e56ffa74985e40882bbd10c295acb15aa104f690bab84de3da37de41a81R2-R7) was an *add*, not a modify 😅

**Comment by trivikr** [comment]
The reason for removing require/import for `react-native-get-random-values` dependency is that our react-native End2End tests fail with "Unable to resolve module" error.

<details>
<summary>Screenshot</summary>

Unable-to-resolve-module-error

</details>

Just like uuid expects it's users to import react-native polyfills, AWS SDK for JavaScript would do the same.

**Comment by trivikr** [comment]
This PR is currently pending review from amplify-js team.
cc @Ashish-Nanda @amhinson

**Comment by trivikr** [comment]
Notes from the meeting on 3/31 with Amplify team:
* Remove requirement of react-native polyfills as peerDependencies, as peerDependenciesMeta just removes the warnings but installs peerDependencies by default.
* Check if a custom error can be emitted for react-native customers requesting particular polyfill as a requirement for aws-sdk just like uuid does https://github.com/uuidjs/uuid#getrandomvalues-not-supported. Amplify team had tried in the past, and it was not straight-forward.

**Review by matrinox — APPROVED** [review]

**Comment by trivikr** [comment]
> * Remove requirement of react-native polyfills as peerDependencies, as peerDependenciesMeta just removes the warnings but installs peerDependencies by default.

Removed in [`5baa663` (#2191)](https://github.com/aws/aws-sdk-js-v3/pull/2191/commits/5baa663adb6b9190c70dc53efd6e0f20d7374c40)

> * Check if a custom error can be emitted for react-native customers requesting particular polyfill as a requirement for aws-sdk just like uuid does [uuidjs/uuid#getrandomvalues-not-supported](https://github.com/uuidjs/uuid#getrandomvalues-not-supported). Amplify team had tried in the past, and it was not straight-forward.

This isn't possible, as it's an implementation detail in https://github.com/uuidjs/uuid/blob/91805f665c38b691ac2cbda56a99231432b00a1a/src/rng-browser.js#L21-L25

I've updated README to share react-native polyfills import example instead [`d67256e` (#2191)](https://github.com/aws/aws-sdk-js-v3/pull/2191/commits/d67256e9d1d07a3e74f3b301efd11f57dadf95ed)

**Review by iartemiev — APPROVED** [review]
Approved after testing the effect of this PR on Amplify RN apps.

**Review by AllanZhengYP — APPROVED** [review]
So nice to get rid of react-native from most users' dependencies.

**Comment by J-Krush** [comment]
So much appreciation for this PR! Unblocked on my work now :D

**Pull request merged**
